File: binance4j-wallet/src/test/java/com/binance4j/wallet/client/WalletClientTest.java
package com.binance4j.wallet.client;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;

import org.junit.jupiter.api.Test;

import com.binance4j.core.exception.ApiException;
import com.binance4j.core.param.TimeFrame;
import com.binance4j.core.test.CustomTest;
import com.binance4j.wallet.dto.DepositStatus;
import com.binance4j.wallet.dto.MarginAccountSnapshotResponse;
import com.binance4j.wallet.dto.WalletTransferType;
import com.binance4j.wallet.dto.WithdrawStatus;
import com.binance4j.wallet.param.AssetDetailParams;
import com.binance4j.wallet.param.AssetDividendRecordParams;
import com.binance4j.wallet.param.DepositAddressParams;
import com.binance4j.wallet.param.DepositHistoryParams;
import com.binance4j.wallet.param.DustTransferParams;
import com.binance4j.wallet.param.FundingWalletParams;
import com.binance4j.wallet.param.TradeFeeParams;
import com.binance4j.wallet.param.WalletTransferHistoryParams;
import com.binance4j.wallet.param.WithdrawHistoryParams;
import com.binance4j.wallet.param.WithdrawParams;
import com.fasterxml.jackson.databind.DeserializationFeature;

public class WalletClientTest extends CustomTest {
	protected WalletClient client = new WalletClient(key, secret);

	@Override
	protected WalletClient getClient() {
		return client;
	}

	@Test
	void testDisableFastWithdrawSwitch() throws ApiException {
		assertDoesNotThrow(() -> getClient().disableFastWithdrawSwitch().sync());
	}

	@Test
	void testEnableFastWithdrawSwitch() throws ApiException {
		assertDoesNotThrow(() -> getClient().enableFastWithdrawSwitch().sync());
	}

	// TODO @Test
	void testdustTransfer() throws ApiException {
		testNotThrow(getClient().dustTransfer(new DustTransferParams(assets)));
	}

	@Test
	void testGetAccountstatus() throws ApiException {
		testNotThrow(getClient().getAccountStatus());
	}

	@Test
	void testGetAllCoinsInfo() throws ApiException {
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		testHasNulls(getClient().getAllCoinsInfo(),
				List.of("country", "specialTips", "specialWithdrawTips", "depositDust"),
				true);
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetApiPermissions() throws ApiException {
		testNotThrow(getClient().getApiPermissions());
	}

	@Test
	void testGetApiTradingStatus() throws ApiException {
		testNotThrow(getClient().getApiTradingStatus());
	}

	@Test
	void testGetAssetDetail() throws ApiException {
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		testNotThrow(getClient().getAssetDetail());
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetAssetDetail2() throws ApiException {
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		testNotThrow(getClient().getAssetDetail(new AssetDetailParams(asset)));
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetAssetDividendRecord() throws ApiException {
		testNotThrow(getClient().getAssetDividendRecord());
	}

	@Test
	void testGetAssetDividendRecord2() throws ApiException {
		testNotThrow(getClient().getAssetDividendRecord(new AssetDividendRecordParams(asset)));
	}

	@Test
	void testGetAssetDividendRecord3() throws ApiException {
		testNotThrow(getClient().getAssetDividendRecord(new AssetDividendRecordParams(asset), new TimeFrame(limit)));
	}

	@Test
	void testGetConvertibleAssets() throws ApiException {
		testNotThrow(getClient().getConvertibleAssets());

	}

	@Test
	void testGetDepositAddress() throws ApiException {
		testNotThrow(getClient().getDepositAddress(new DepositAddressParams("BNB")));
	}

	@Test
	void testGetDepositHistory() throws ApiException {
		testNotThrow(getClient().getDepositHistory());
	}

	@Test
	void testGetDepositHistory2() throws ApiException {
		testNotThrow(getClient().getDepositHistory(new DepositHistoryParams(asset)));
	}

	@Test
	void testGetDepositHistory3() throws ApiException {
		testNotThrow(getClient().getDepositHistory(new DepositHistoryParams(DepositStatus.SUCCESS)));
	}

	@Test
	void testGetDepositHistory4() throws ApiException {
		testNotThrow(getClient().getDepositHistory(new DepositHistoryParams(DepositStatus.SUCCESS, asset)));
	}

	@Test
	void testGetDustLog() throws ApiException {
		testNotThrow(getClient().getDustLog());

	}

	@Test
	void testGetDustLog2() throws ApiException {
		testNotThrow(getClient().getDustLog(new TimeFrame(10)));

	}

	@Test
	void testgetFundingWallet() throws ApiException {
		testNotThrow(getClient().getFundingWallet());
	}

	@Test
	void testgetFundingWallet2() throws ApiException {
		testNotThrow(getClient().getFundingWallet(new FundingWalletParams(asset)));
	}

	@Test
	void testgetFundingWallet3() throws ApiException {
		testNotThrow(getClient().getFundingWallet(new FundingWalletParams(asset, true)));
	}

	@Test
	void testGetFuturesAccountSnapshot() throws ApiException {
		testNotThrow(getClient().getFuturesAccountSnapshot());
	}

	@Test
	void testGetFuturesAccountSnapshot2() throws ApiException {
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		testNotThrow(getClient().getFuturesAccountSnapshot(new TimeFrame(limit)));
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetMarginAccountSnapshot() throws ApiException {
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		MarginAccountSnapshotResponse res = getClient().getMarginAccountSnapshot(new TimeFrame(limit)).sync();
		assertDoesNotThrow(() -> assertTrue(getNullProperties(res, true).contains("marginLevel")));
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetSpotAccountSnapshot() throws ApiException {
		testNotThrow(getClient().getSpotAccountSnapshot());
	}

	@Test
	void testGetSpotAccountSnapshot2() throws ApiException {
		testNotThrow(getClient().getSpotAccountSnapshot(new TimeFrame(limit)));
	}

	@Test
	void testGetSystemStatus() throws ApiException {
		testNotThrow(getClient().getSystemStatus());
	}

	@Test
	void testGetTradeFee() throws ApiException {
		testNotThrow(getClient().getTradeFee());
	}

	@Test
	void testGetTradeFee2() throws ApiException {
		testNotThrow(getClient().getTradeFee(new TradeFeeParams("BNBBTC")));
	}

	@Test
	void testGetTransferHistory() throws ApiException {
		testNotThrow(getClient().getTransferHistory(new WalletTransferHistoryParams(WalletTransferType.MAIN_MARGIN)));
	}

	@Test
	void testGetTransferHistory2() throws ApiException {
		testNotThrow(client
				.getTransferHistory(new WalletTransferHistoryParams(WalletTransferType.MAIN_MARGIN, asset, asset)));
	}

	@Test
	void testGetWithdrawHistory() throws ApiException {
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		testNotThrow(getClient().getWithdrawHistory());
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);
	}

	@Test
	void testGetWithdrawHistory2() throws ApiException {
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false);
		WithdrawStatus status = WithdrawStatus.COMPLETED;
		testNotThrow(getClient().getWithdrawHistory(new WithdrawHistoryParams("FTM", status)));
		getClient().getMapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true);

	}

	@Test
	void testTransfer() throws ApiException {

	}

	// @Test
	void testWithdraw() throws ApiException {
		testNotThrow(getClient().withdraw(new WithdrawParams("amount", "coin", "address")));
	}
}

File: connectors/src/test/kotlin/com/binance4j/wallet/client/WalletClientTest.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.binance4j.wallet.client

import com.binance4j.core.client.RestClient
import com.binance4j.core.exception.ApiException
import com.binance4j.core.param.TimeFrame
import com.binance4j.core.test.CustomTest
import com.binance4j.wallet.dto.DepositStatus
import com.binance4j.wallet.dto.WalletTransferType
import com.binance4j.wallet.dto.WithdrawStatus
import com.binance4j.wallet.param.*
import com.fasterxml.jackson.databind.DeserializationFeature
import org.junit.jupiter.api.Test

class WalletClientTest : CustomTest() {
    private var client = WalletClient(key, secret)

    @Test
    @Throws(ApiException::class)
    fun testDisableFastWithdrawSwitch() {
        assertNotThrow(client.disableFastWithdrawSwitch())
    }

    @Test
    @Throws(ApiException::class)
    fun testEnableFastWithdrawSwitch() {
        assertNotThrow(client.enableFastWithdrawSwitch())
    }

    // TODO @Test
    @Throws(ApiException::class)
    fun testdustTransfer() {
        assertNotThrow(client.dustTransfer(DustTransferParams(assets)))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetAccountstatus() {
        assertNotThrow(client.getAccountStatus())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetAllCoinsInfo() {
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false)
        assertNotThrow(client.getAllCoinsInfo())
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
    }

    @Test
    @Throws(ApiException::class)
    fun testGetApiPermissions() {
        assertNotThrow(client.getApiPermissions())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetApiTradingStatus() {
        assertNotThrow(client.getApiTradingStatus())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetAssetDetail() {
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false)
        assertNotThrow(client.getAssetDetail())
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
    }

    @Test
    @Throws(ApiException::class)
    fun testGetAssetDetail2() {
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false)
        assertNotThrow(client.getAssetDetail(AssetDetailParams(asset)))
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
    }

    @Test
    @Throws(ApiException::class)
    fun testGetAssetDividendRecord() {
        assertNotThrow(client.getAssetDividendRecord())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetAssetDividendRecord2() {
        assertNotThrow(client.getAssetDividendRecord(AssetDividendRecordParams(asset)))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetAssetDividendRecord3() {
        assertNotThrow(client.getAssetDividendRecord(AssetDividendRecordParams(asset), TimeFrame(limit.toLong())))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetConvertibleAssets() {
        assertNotThrow(client.getConvertibleAssets())
    }

    // @Test OK but restricted on GitHub Actions (USA?)
    @Throws(ApiException::class)
    fun testGetDepositAddress() {
        assertNotThrow(client.getDepositAddress(DepositAddressParams("BNB")))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetDepositHistory() {
        assertNotThrow(client.getDepositHistory())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetDepositHistory2() {
        assertNotThrow(client.getDepositHistory(DepositHistoryParams(asset)))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetDepositHistory3() {
        assertNotThrow(client.getDepositHistory(DepositHistoryParams(DepositStatus.SUCCESS)))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetDepositHistory4() {
        assertNotThrow(client.getDepositHistory(DepositHistoryParams(asset, DepositStatus.SUCCESS)))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetDustLog() {
        assertNotThrow(client.getDustLog())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetDustLog2() {
        assertNotThrow(client.getDustLog(TimeFrame(10)))
    }

    @Test
    @Throws(ApiException::class)
    fun testgetFundingWallet() {
        assertNotThrow(client.getFundingWallet())
    }

    @Test
    @Throws(ApiException::class)
    fun testgetFundingWallet2() {
        assertNotThrow(client.getFundingWallet(FundingWalletParams(asset)))
    }

    @Test
    @Throws(ApiException::class)
    fun testgetFundingWallet3() {
        assertNotThrow(client.getFundingWallet(FundingWalletParams(asset, true)))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetFuturesAccountSnapshot() {
        assertNotThrow(client.getFuturesAccountSnapshot())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetFuturesAccountSnapshot2() {
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false)
        assertNotThrow(client.getFuturesAccountSnapshot(TimeFrame(limit.toLong())))
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
    }

    @Test
    @Throws(ApiException::class)
    fun testGetMarginAccountSnapshot() {
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false)
        assertNotThrow(client.getMarginAccountSnapshot(TimeFrame(limit.toLong())))
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
    }

    @Test
    @Throws(ApiException::class)
    fun testGetSpotAccountSnapshot() {
        assertNotThrow(client.getSpotAccountSnapshot())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetSpotAccountSnapshot2() {
        assertNotThrow(client.getSpotAccountSnapshot(TimeFrame(limit.toLong())))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetSystemStatus() {
        assertNotThrow(client.getSystemStatus())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetTradeFee() {
        assertNotThrow(client.getTradeFee())
    }

    @Test
    @Throws(ApiException::class)
    fun testGetTradeFee2() {
        assertNotThrow(client.getTradeFee(TradeFeeParams("BNBBTC")))
    }

    @Test
    @Throws(ApiException::class)
    fun testGetTransferHistory() {
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false)
        assertNotThrow(client.getTransferHistory(WalletTransferHistoryParams(WalletTransferType.MAIN_MARGIN)))
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
    }

    @Test
    @Throws(ApiException::class)
    fun testGetTransferHistory2() {
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false)
        assertNotThrow(
            client.getTransferHistory(WalletTransferHistoryParams(WalletTransferType.MAIN_MARGIN, asset, asset))
        )
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
    }

    @Test
    @Throws(ApiException::class)
    fun testGetWithdrawHistory() {
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false)
        assertNotThrow(client.getWithdrawHistory())
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
    }

    @Test
    @Throws(ApiException::class)
    fun testGetWithdrawHistory2() {
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, false)
        assertNotThrow(client.getWithdrawHistory(WithdrawHistoryParams("FTM", null, WithdrawStatus.COMPLETED)))
        RestClient.mapper().configure(DeserializationFeature.FAIL_ON_MISSING_CREATOR_PROPERTIES, true)
    }

    @Test
    @Throws(ApiException::class)
    fun testTransfer() {
    }

    // @Test
    @Throws(ApiException::class)
    fun testWithdraw() {
        assertNotThrow(client.withdraw(WithdrawParams("amount", "coin", "address")))
    }
}